lib/ansible/module_utils/basic.py

```diff
@@ -411,7 +416,10 @@ def remove_values(value, no_log_strings):
         old_data, new_data = deferred_removals.popleft()
         if isinstance(new_data, Mapping):
             for old_key, old_elem in old_data.items():
-                new_key = _remove_values_conditions(old_key, no_log_strings, deferred_removals)
+                if old_key in _NO_MODIFY_KEYS or old_key.startswith('_ansible'):
+                    new_key = old_key
+                else:
+                    new_key = _remove_values_conditions(old_key, no_log_strings, deferred_removals)
```

Comment:
I don't understand the motivation behind stripping keys at all, I feel like this can conditionally break the contract defined by the `RETURN` structure of modules as documented. Is there a nuance I'm missing as to why that's desired/necessary?

Reply:
As described in the previous issue, we were made aware that modules could return no_log values in keys.  In the example of the `uri` module, we have no control over the data format returned by the API, and as such, `no_log` data can appear in a key. Let's say that a module marks `user` as `no_log`, and the return of the module looks like:

We decided this was a security vulnerability.

That decision is not likely to be backed out.  However, what this attempts to achieve, is that this change doesn't step on assumptions by the ansible code base, specifically around whether or not a task can show up as changed/etc in the output.

Reply:
This skips _modifying_ certain keys, not returning them. If a key contains a string that is in `no_log_strings`, we mask the value in the key not just the value.

Reply:
Previously we only masked the value in the key, which could result in information disclosure if a `no_log` value ended up in they key.

Reply:
You are revealing the no_log value with that. The missing part in the key is the password.

Reply:
Wait, so the original issue that was defined as a security vulnerability is that modules are exposing `no_log` data as a *KEY* in the past?

Reply:
Also, if we're masking the string value of a *key* we're effectively breaking the data contract of the `RETURN` structure which will 100% break playbooks (as we've found out). This seems like a critical bug to me, I would instead suggest that if a key contains a `no_log` string then we should consider the module as failed with an error message as to why because it should be considered a bug in the module.

Reply:
@maxamillion Yes, the `uri` module was getting a response back from an API that had username/passwords in the keys in the response (it was a substring of the key, not the entire key).

Reply:
@samdoran that sounds like a but in the `uri` module ... is it possible for the introspection of the return data to check the `no_log` string and sanitize it's data in the module or error and provide a reasonable message?

I fear that swinging the hammer on the controller side like this is too heavy handed and will break a considerable amount of playbooks on accident. Granted, it might only be for CI or testing infrastructure were passwords are weaker and more likely that `no_log` data would accidentally match a key value but it's still going to be a game of whack-a-mole to keep adding "special keys" (I see this going sideways in Collections land). What I would propose is that a `no_log` value in a key is considered an error and the controller should fail the task with an appropriate error message and then module developers should handle this on their end, if possible.

Reply:
@maxamillion I don't know that your idea of an error would be any more palatable to those involved here.  While right now you at least get something back, it would prevent the module from working at all for affected users.

Reply:
@maxamillion The issue is that I do not have control over a password a user chooses. If the user is selecting a password that is completely or partly matching a parameter that a module is returning, the module will not be usable any more. An error would result in the same issue as now. It would not be predictable.

We have roles in ansible-freeipa that are using lots of modules. I am not sure that a role or module could compare a sensitive user password to all the returned parameter names to fail early before even doing changes.

In my opinion it is the responsibility of a module to make sure that no sensitive information is leaked into the returned parameters. I do not see a way how to address this in Ansible without introducing issues like https://github.com/ansible/ansible/issues/70455. Ansible does simply not know if there is an overlap of sensitive and insensitive information only because a user has chosen a password that accidentally matches the name or parts of the name of a parameter.

Reply:
@t-woerner it is the responsibility for the module to not leak information, but there have been CVEs against Ansible because `no_log` unintentionally leaked information because that responsibility was not upheld, so now there is oversight to attempt to ensure that it won't happen even if there's a module bug. We have to define a solution that allows that oversight to continue to function in a reasonable way. I'm open to options here but I think because of the history of CVE disclosures about this the Ansible dev community are large are sensitive to the issue.